dll/win32/shell32/shlexec.cpp

```diff
@@ -1983,21 +1983,43 @@ static BOOL SHELL_execute(LPSHELLEXECUTEINFOW sei, SHELL_ExecuteW32 execfunc)
     else
         *wszParameters = L'\0';
 
+    // Get the working directory
     WCHAR dirBuffer[MAX_PATH];
     LPWSTR wszDir = dirBuffer;
+    ::GetCurrentDirectoryW(_countof(dirBuffer), dirBuffer);
     CHeapPtr<WCHAR, CLocalAllocator> wszDirAlloc;
-    if (sei_tmp.lpDirectory)
+    if (sei_tmp.lpDirectory && *sei_tmp.lpDirectory)
     {
-        len = lstrlenW(sei_tmp.lpDirectory) + 1;
-        if (len > _countof(dirBuffer))
+        if (sei_tmp.fMask & SEE_MASK_DOENVSUBST)
         {
-            wszDirAlloc.Allocate(len);
-            wszDir = wszDirAlloc;
+            LPWSTR tmp = expand_environment(sei_tmp.lpDirectory);
+            if (tmp)
+            {
+                wszDirAlloc.Attach(tmp);
+                wszDir = wszDirAlloc;
+            }
+        }
+        else
+        {
+            __SHCloneStrW(&wszDirAlloc, sei_tmp.lpDirectory);
+            if (wszDirAlloc)
+                wszDir = wszDirAlloc;
+        }
+    }
+    // NOTE: ShellExecute should accept the invalid working directory
+    if (!PathIsDirectoryW(wszDir) || lstrcmpiW(wszDir, L".\\") == 0)
```

Comment:
What about ".", "\\" or ".\\foo" or all the other relative cases?

Reply:
`".\\"` is a special case. The others are not.

Reply:
```suggestion
    // NOTE: ".\\" is a special case
    if (!PathIsDirectoryW(wszDir) || lstrcmpiW(wszDir, L".\\") == 0)
```